src/write_syslog.c

```diff
@@ -0,0 +1,666 @@
+    } else if (((size_t)status) >= (ret_len - offset)) {                       \
+      sfree(rates);                                                            \
+      return -1;                                                               \
+    } else                                                                     \
+      offset += ((size_t)status);                                              \
+  } while (0)
+
+  if (ds->ds[ds_num].type == DS_TYPE_GAUGE)
+    BUFFER_ADD(GAUGE_FORMAT, vl->values[ds_num].gauge);
+  else if (store_rates) {
+    if (rates == NULL)
+      rates = uc_get_rate(ds, vl);
+    if (rates == NULL) {
+      WARNING("format_values: "
+              "uc_get_rate failed.");
+      return -1;
+    }
+    BUFFER_ADD(GAUGE_FORMAT, rates[ds_num]);
+  } else if (ds->ds[ds_num].type == DS_TYPE_COUNTER)
+    BUFFER_ADD("%" PRIu64, (uint64_t)vl->values[ds_num].counter);
+  else if (ds->ds[ds_num].type == DS_TYPE_DERIVE)
+    BUFFER_ADD("%" PRIi64, vl->values[ds_num].derive);
+  else if (ds->ds[ds_num].type == DS_TYPE_ABSOLUTE)
+    BUFFER_ADD("%" PRIu64, vl->values[ds_num].absolute);
+  else {
+    ERROR("format_values plugin: Unknown data source type: %i",
+          ds->ds[ds_num].type);
+    sfree(rates);
+    return -1;
+  }
+
+#undef BUFFER_ADD
+
+  sfree(rates);
+  return 0;
+}
+
+static int ws_format_name(char *ret, int ret_len, const value_list_t *vl,
+                          const struct ws_callback *cb, const char *ds_name) {
+
+  if (ds_name != NULL) {
+    if (vl->type[0] != '\0') {
+      snprintf(ret, ret_len, "%s.%s", vl->type, ds_name);
+    }
+  } else { /* ds_name == NULL */
+    if (vl->type[0] != '\0') {
+      snprintf(ret, ret_len, "%s", vl->type);
+    }
+  }
+
+  return 0;
+}
+
+static int ws_send_message(const char *key, const char *value, cdtime_t time,
+                           struct ws_callback *cb, const char *plugin,
+                           const char *plugin_instance,
+                           const char *type_instance, const char *type,
+                           const char *ds_name, cdtime_t interval,
+                           const char *host) {
+  int status;
+  size_t message_len;
+  char message[1024];
+  char *buffer = "";
+  const char *host_tags = cb->host_tags ? cb->host_tags : "";
+  const char *host_tags_json_prefix = "";
+  const char *metrics_prefix =
+      cb->metrics_prefix ? cb->metrics_prefix : WS_DEFAULT_PREFIX;
+  const char *msg_format = cb->msg_format ? cb->msg_format : WS_DEFAULT_FORMAT;
+
+  /* skip if value is NaN */
+  if (value[0] == 'n')
+    return 0;
+
+  if (strcasecmp("JSON", msg_format) == 0) {
+    if (cb->host_tags) {
+      host_tags_json_prefix = ",";
+    }
+    status = snprintf(
+        /* The metric key-values are are part of the syslog msg, in json
+           format */
+        message, sizeof(message),
+        "<166>%s collectd {\"time\":%.0f, \"%s\":{ \"%s\":{ \"%s\":%s }, "
```

Comment:
Could you add a pointer to docs describing this format?

Reply:
How should that be done?

Reply:
so the concerns I have with the proposed json structure are that some of the json keys are unpredictable: they contain the value of `type_instance`. I'd rather have `{"type_instance":"foobarbaz","value":1234}` than `{"foobarbaz":1234}`. The latter will for instance create an unpredictable number of fields in Elasticsearch.

Reply:
I meant a comment with a link to the RFC, doc, or – failing all that – the implementation.

Reply:
I don't have an RFC. Added explanation in the collectd.conf.pod.